group/assignment6/RearrangeCars.cpp

```diff
@@ -0,0 +1,221 @@
+            bool foundMisplacedCar = false;
+            for (int carId = 0; carId < carPosition.size(); carId++) {
+                  if (lastMovedCarId != carId && endStates[carPosition[carId]] != carId &&
+                      validMove(carId, emptySlotId, constraints)) {
+                        currCarSlotId = carPosition[carId];
+                        lastMovedCarId = carId;
+
+                        foundMisplacedCar = true;
+                        break;
+                  }
+            }
+
+            // If misplaced car can't be moved to free slot, move a car that is on its position and can be moved here
+            if (!foundMisplacedCar) {
+                  for (int carId = 0; carId < carPosition.size(); carId++) {
+                        if (validMove(carId, emptySlotId, constraints)) {
+                              currCarSlotId = carPosition[carId];
+                              lastMovedCarId = carId;
+
+                              numOfCarsOnDisiredPosition--;
+                              break;
+                        }
+                  }
+            }
+
+            // Make a move
+            moveCar(currentStates, carPosition, currCarSlotId, emptySlotId);
+            sequenceOfMoves.push_back(Move(lastMovedCarId, make_pair(currCarSlotId, emptySlotId)));
+            emptySlotId = currCarSlotId;
+      }
+
+}
+
+bool RearrangeCars::generateSequenceOfMoves(const vector<int> &endStates, vector<Move> &sequenceOfMoves,
+                                            const vector<set<int>> *constraints) {
+      sequenceOfMoves.clear();
+      if (!isInputValid(endStates, constraints)) {
+            return false;
+      }
+      vector<int> currentStates = this->startStates; // startStates will remain unchanged
+      vector<int> carPosition = findCarPositions(currentStates);
+
+      int emptySlotId = findEmptySlotId();
+      int currCarId, currCarSlotId;
+      int numOfCarsOnDisiredPosition = numberOfCarsOnTheirDesiredPositions(carPosition, endStates);
+
+      while (true) {
+            currCarId = endStates[emptySlotId];
+            if (currCarId == EMPTY_SLOT_ID) // We have reached slot which should be empty in the end
+                  currCarId = findMisplacedCar(carPosition, currentStates, endStates, constraints,
+                                               sequenceOfMoves, emptySlotId, numOfCarsOnDisiredPosition);
+
+            currCarSlotId = carPosition[currCarId];
+            moveCar(currentStates, carPosition, currCarSlotId, emptySlotId);
+            numOfCarsOnDisiredPosition++;
+            sequenceOfMoves.push_back(Move(currCarId, make_pair(currCarSlotId, emptySlotId)));
+            emptySlotId = currCarSlotId;
+
+            if (numOfCarsOnDisiredPosition == carPosition.size())
+                  break;
+      }
+      return true;
+}
+
+void
+RearrangeCars::bruteForceAllSequence(vector<int> &currentStates, const vector<int> &endStates, vector<int> &carPosition,

```

Comment:
Must be renamed!

Reply:
I agree. Also sequenceOfMoves and sequencesOfMoves are very similar to each other so it's easy to get confused here. Maybe we should rename arguments, too?

In my opinion method might be called "getAllSequencesFromCurPosition", and arguments might be "sequenceOfMoves" and "allSequencesOfMoves", what do you think?

Reply:
I like it. We can also call it "getAllSequencesFromCur*rent*Position" if we're using a long name anyways..
I also agree about the variable names

Also about variable names, I said it before but nobody said what she thinks.. I think startStates, endStates, carPosition are very confusing name.. I would call them currentSlot2CarId, desiredSlot2CarId, currentCarId2Slot. Or something like that.. What do you think?

Reply:
Oh yes, I just wanted to comment about Current. Sometimes we shorten it to Curr,  sometimes we use the full word. I think both are ok, but my mistake here was to shorten it to Cur.

Well... For me startStates, endStates and carPositions are pretty understandable, but maybe it's because I got used to them during this assignment. I think your version is better :+1:  If everyone agrees, we can fix that.

Reply:
I fixed startStates and endStates names, but kept carPositions. For me introducing CarId2Slot may be a bit confusing for thre reader, but I am ready to discuss.